pkg/agent/evictionmanager/plugin/memory/enhanced_numa_pressure.go

```diff
@@ -0,0 +1,305 @@
+		return nil, fmt.Errorf("GetTopEvictionPods got nil request")
+	}
+
+	if len(request.ActivePods) == 0 {
+		general.Warningf("GetTopEvictionPods got empty active pods list")
+		return &pluginapi.GetTopEvictionPodsResponse{}, nil
+	}
+
+	dynamicConfig := m.dynamicConfig.GetDynamicConfiguration()
+	targetPods := make([]*v1.Pod, 0, len(request.ActivePods))
+	podToEvictList := make([]*v1.Pod, 0, len(request.ActivePods))
+	podToEvictMap := make(map[string]*v1.Pod)
+
+	general.Infof("GetTopEvictionPods condition, isUnderNumaPressure: %+v, m.numaActionMap: %+v",
+		m.isUnderNumaPressure,
+		m.numaActionMap)
+
+	if dynamicConfig.EnableEnhancedNumaLevelEviction && m.isUnderNumaPressure {
+		for numaID, action := range m.numaActionMap {
+			candidates := m.getCandidates(request.ActivePods, numaID, dynamicConfig.NumaVictimMinimumUtilizationThreshold)
+			m.evictionHelper.selectTopNPodsToEvictByMetrics(candidates, request.TopN, numaID, action,
+				dynamicConfig.NumaEvictionRankingMetrics, podToEvictMap)
+		}
+	}
+
+	for uid := range podToEvictMap {
+		podToEvictList = append(podToEvictList, podToEvictMap[uid])
+	}
+	targetPods = m.getTargets(podToEvictList, targetPods)
+
+	_ = m.emitter.StoreInt64(metricsNameNumberOfTargetPods, int64(len(targetPods)), metrics.MetricTypeNameRaw)
+	general.Infof("[enhanced-numa-memory-pressure-eviction-plugin] GetTopEvictionPods result, targetPods: %+v", native.GetNamespacedNameListFromSlice(targetPods))
+
+	resp := &pluginapi.GetTopEvictionPodsResponse{
+		TargetPods: targetPods,
+	}
+	if gracePeriod := dynamicConfig.MemoryPressureEvictionConfiguration.GracePeriod; gracePeriod > 0 {
+		resp.DeletionOptions = &pluginapi.DeletionOptions{
+			GracePeriodSeconds: gracePeriod,
+		}
+	}
+
+	return resp, nil
+}
+
+func (m *EnhancedNumaMemoryPressurePlugin) getTargets(podToEvictList []*v1.Pod, targetPods []*v1.Pod) []*v1.Pod {
+	predictNotUnderPressure := func(predictNumaPressureStatMap map[int]float64) bool {
+		for _, numaAvailMemSubtractWatermark := range predictNumaPressureStatMap {
+			if numaAvailMemSubtractWatermark <= 0 {
+				return false
+			}
+		}
+		return true
+	}
+	releaseNumaPressureByEvictPod := func(predictNumaPressureStatMap map[int]float64, pod *v1.Pod) {
+		for numaID := range predictNumaPressureStatMap {
+			podNumaTotal, err := helper.GetPodMetric(m.metaServer.MetricsFetcher, m.emitter, pod, consts.MetricsMemTotalPerNumaContainer, numaID)
+			if err != nil {
+				general.Infof("Failed to get pod  %+v total mem usage on numa id %+v", pod.UID, numaID)
+				continue
+			}
+			// TODO: it's better to use sum of podNumaAnon and podNumaActiveFile to predict how much the pod can release the pressure
+			//  since inactive file is already counted in predictNumaPressureStatMap, while cgroupv1 does not provide numa active/inactive statistics.
+			//  So we falls back to use pod numa total.
+			predictNumaPressureStatMap[numaID] += podNumaTotal
+		}
+	}
+	getPodPermutations := func(podList []*v1.Pod) [][]*v1.Pod {
```

Comment:
when `podList` get larger, this function will be costly, is it really necessary?